edgelesssys/constellation
# aws: use new LB controller to fix SecurityGroup cleanup on K8s service deletion

### Context

The AWS cloud controller manager (CCM), which is currently responsible for LB creation, has issues on cleaning up security groups on K8s service deletion.  
This leads to problems of `constellation terminate` (see #2105).
We addressed this issue upstream (see https://github.com/kubernetes/cloud-provider-aws/issues/566), but did not get any help.

There is a newer controller called `aws-load-balancer-controller` which has some responsibility overlap with the CCM, but its distinction is unfortunately not well documented (https://github.com/kubernetes/cloud-provider-aws/issues/203).
The official docs shortly mention that its usage is preferred over CCM, which is not actively developed anymore (see https://docs.aws.amazon.com/eks/latest/userguide/network-load-balancing.html#:~:text=We%20recommend%20that%20you%20use,the%20AWS%20Load%20Balancer%20Controller.).

### Proposed change(s)
IMPORTANT: we disable the iam upgrade because of permission issues and discuss a solution on how to best upgrade iam permissions. The upgrade still works when the user manually sets the new permissions



- use the newer, dedicated `aws-load-balancer-controller` for managing the LBs
  - disable the LB service from the CCM
  - on `create` install the a new helm chart `aws-load-balancer-controller` which needs to be installed after `certManager`
  - add new IAM permissions for the new controller
- on `upgrade apply`
  - apply updated IAM profile
  - install `aws-load-balancer-controller` if not yet installed, otherwise upgrade 
 - document the AWS external LB service creation


~TODO~: it's not easily visible when the DNS endpoint from the LB is reachable. The K8s service publishes the endpoint way before the LB is provisioned. How to check that the service is externally reachable?

-> Amazon suggests to go the AWS portal and wait for the LB targets to be ready before pinging the DNS endpoint. So our current doc seems ok. https://docs.aws.amazon.com/eks/latest/userguide/network-load-balancing.html#:~:text=Wait%20until%20the%20status%20of%20all%20targets%20is%20healthy%20before%20continuing.%20It%20might%20take%20several%20minutes%20before%20all%20targets%20are%20healthy.

~TODO~: how to deal with .helmignore issue? Our implementation is unable to apply rules to files in subdirectories (e.g. `crds/kustomization.yaml`); But the fix from upstream in the helm sdk (using `symwalk`) causes problems when loading cilium in production (no idea why, see TODO in code)

-> keep current implementation and just add an annotation to the function about the caveat

### How to test
- check upgrade (iam roles and terraform resources should update, + new aws-lb chart)
- lb-test should work
- delete lb service, then delete the cluster



### Additional info
[E2E lb run](https://github.com/edgelesssys/constellation/actions/runs/5590440541); new run after changes https://github.com/edgelesssys/constellation/actions/runs/5629928898/job/15255421418 ✅
[E2e upgrade run](https://github.com/edgelesssys/constellation/actions/runs/5607978039/jobs/10259768263) ✅ from v2.9.0; tested manually locally after changes

co-authored by @malt3 

### Checklist
